### PR TITLE
Bugfix for nodata handling in `extract_strahler_streams_d8`

### LINKED ISSUE #363: calculate_subwatershed_boundary TypeError given a NoneType argument
```
(C:\Users\dmf\projects\invest\env) λ invest -vvv run routedem -d data\invest-sample-data\RouteDEM\routedem_gura.invs.json -w ../scratch/routedem2
12/11/2023 16:32:17  natcap.invest.cli  INFO     Imported target natcap.invest.routedem from <module 'natcap.invest.routedem' from 'C:\\Users\\dmf\\projects\\invest\\env\\lib\\site-packages\\natcap\\invest\\routedem.py'>
12/11/2023 16:32:17  natcap.invest.utils Level 100 Writing log messages to [../scratch/routedem2\InVEST-natcap.invest.routedem-log-2023-12-11--16_32_17.txt]
12/11/2023 16:32:17  natcap.invest.cli  Level 100 Starting model with parameters:
Arguments for InVEST natcap.invest.routedem 3.13.0.post775+gf7597fa27:
algorithm                     D8
calculate_downstream_distance False
calculate_flow_accumulation   True
calculate_flow_direction      True
calculate_slope               True
calculate_stream_order        True
calculate_stream_threshold    True
calculate_subwatersheds       True
dem_band_index
dem_path                      C:\Users\dmf\projects\invest\data\invest-sample-data\RouteDEM\DEM_gura.tif
results_suffix                gura
threshold_flow_accumulation   1000
workspace_dir                 ../scratch/routedem2

12/11/2023 16:32:17  natcap.invest.routedem INFO     Using DEM band index 1
12/11/2023 16:32:18  pygeoprocessing.routing.routing INFO     (fill pits): complete
12/11/2023 16:32:18  natcap.invest.routedem INFO     calculating flow direction
12/11/2023 16:32:18  pygeoprocessing.routing.routing INFO     (flow dir d8): complete
12/11/2023 16:32:18  natcap.invest.routedem INFO     calculating flow accumulation
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     Flow accumulation D8 100.0% complete
12/11/2023 16:32:19  pygeoprocessing.geoprocessing INFO     starting stats_worker
12/11/2023 16:32:19  pygeoprocessing.geoprocessing_core DEBUG    stats worker PID: 3188
12/11/2023 16:32:19  pygeoprocessing.geoprocessing INFO     started stats_worker <Thread(Thread-1 (stats_worker), started daemon 21176)>
12/11/2023 16:32:19  pygeoprocessing.geoprocessing INFO     100.0% complete
12/11/2023 16:32:19  pygeoprocessing.geoprocessing INFO     Waiting for raster stats worker result.
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): seed the drains
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): drain seeding complete
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): starting upstream walk
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): stream segment creation complete
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): determining stream order
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): stream order processing complete
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): determine rivers
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): flow accumulation adjustment complete
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): final pass on stream order and geometry
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): final pass on stream order complete
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): commit transaction due to stream joining
12/11/2023 16:32:19  pygeoprocessing.routing.routing INFO     (extract_strahler_streams_d8): all done
12/11/2023 16:32:20  taskgraph.Task     ERROR    Something went wrong when adding task Calculate subwatersheds from stream order (6), terminating taskgraph.
Traceback (most recent call last):
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 674, in add_task
    new_task._call()
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 1093, in _call
    payload = self._func(*self._args, **self._kwargs)
  File "src\pygeoprocessing\routing\routing.pyx", line 4099, in pygeoprocessing.routing.routing.calculate_subwatershed_boundary
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
12/11/2023 16:32:20  natcap.invest.utils ERROR    Exception while executing natcap.invest.routedem
Traceback (most recent call last):
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\natcap\invest\utils.py", line 164, in prepare_workspace
    yield
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\natcap\invest\cli.py", line 470, in main
    model_module.execute(parsed_datastack.args)
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\natcap\invest\routedem.py", line 485, in execute
    graph.add_task(
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 674, in add_task
    new_task._call()
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 1093, in _call
    payload = self._func(*self._args, **self._kwargs)
  File "src\pygeoprocessing\routing\routing.pyx", line 4099, in pygeoprocessing.routing.routing.calculate_subwatershed_boundary
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
12/11/2023 16:32:20  natcap.invest.utils INFO     Elapsed time: 2.31s
12/11/2023 16:32:20  natcap.invest.utils INFO     Execution finished
Traceback (most recent call last):
  File "C:\Users\dmf\projects\invest\env\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\dmf\projects\invest\env\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\dmf\projects\invest\env\Scripts\invest.exe\__main__.py", line 7, in <module>
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\natcap\invest\cli.py", line 470, in main
    model_module.execute(parsed_datastack.args)
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\natcap\invest\routedem.py", line 485, in execute
    graph.add_task(
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 674, in add_task
    new_task._call()
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\taskgraph\Task.py", line 1093, in _call
    payload = self._func(*self._args, **self._kwargs)
  File "src\pygeoprocessing\routing\routing.pyx", line 4099, in pygeoprocessing.routing.routing.calculate_subwatershed_boundary
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

### PR DESCRIPTION
**Pull request by davemfish**
Fixes #363 

We needed a `double` instead of a `long` to accommodate the nodata value of a flow_accumulation raster. And we needed to check for the presence of that nodata value.

I modified some test data to raise the exception seen in #363 before applying this fix. But it's such a special construction of the DEM that it makes the expected values under test (number of stream features, number of watersheds, etc) much harder to reason about. After doing all that, I think it might actually be better to leave the test as it was, instead of manipulating it to cover this edge case. What do you think?

### CODE REVIEW

File: src/pygeoprocessing/routing/routing.pyx [hunk 1]
```diff
@@ -3354,16 +3355,17 @@ def extract_strahler_streams_d8(
...
+                        _is_close(local_flow_accum,
+                                  flow_accum_nodata, 1e-8, 1e-5)):
                     continue
                 # check to see if it's a drain
                 d_n = <int>flow_dir_managed_raster.get(x_l, y_l)
                 x_n = x_l + D8_XOFFSET[d_n]
                 y_n = y_l + D8_YOFFSET[d_n]
-
-                if (x_n < 0 or y_n < 0 or x_n >= n_cols or y_n >= n_cols or
+                if (x_n < 0 or y_n < 0 or x_n >= n_cols or y_n >= n_rows or
```
Comment (davemfish) [inline src/pygeoprocessing/routing/routing.pyx:3368]: This was an unrelated bug, `y_n >= n_rows` is the fix. A similar subsequent check for this meant that this bug never resulted in any bad results, only non-optimal searches. Given that, I didn't see a reasonable way to test this.

Comment (phargogh) [reply src/pygeoprocessing/routing/routing.pyx:3368]: Sounds ok to me!  There are a number of things that have come up in pygeoprocessing's routing over the years that are subtle bugs that are difficult to test or required impractically large test data.  In both these cases, we just patched them and didn't worry about the tests.  I'm a-ok with continuing this practice if you are!

File: src/pygeoprocessing/routing/routing.pyx [hunk 2]
```diff
@@ -3354,16 +3355,17 @@ def extract_strahler_streams_d8(
                 is_drain = 0
                 x_l = xoff + i
                 y_l = yoff + j
-                local_flow_accum = <long>flow_accum_managed_raster.get(
+                local_flow_accum = <double>flow_accum_managed_raster.get(
```
Comment (phargogh) [inline src/pygeoprocessing/routing/routing.pyx:3358]: Strictly speaking the `<double>` probably isn't needed since `.get()` already returns a double.  But it also isn't hurting anything by leaving it, and the compiler might well optimize it out anyways!

File: src/pygeoprocessing/routing/routing.pyx [hunk 3]
```diff
@@ -4871,7 +4877,7 @@ cdef _calculate_stream_geometry(
 def _delete_feature(
         stream_feature, stream_layer, upstream_to_downstream_id,
         downstream_to_upstream_ids):
-    """Helper for Mahler extraction to delete all references to a stream.
```
Comment (phargogh) [inline src/pygeoprocessing/routing/routing.pyx:4874]: Ha, can't believe I didn't catch this earlier!

**Review by phargogh — APPROVED** [review]
The one failing test is on building a mac wheel, and the job was just cancelled.  Everything looks great, thanks for digging into this @davemfish !
